--- a/node_modules/powerbi-visuals-tools/lib/VisualBuilder.js	2017-12-19 12:55:33.000000000 +0100
+++ b/node_modules/powerbi-visuals-tools/lib/VisualBuilder.js	2018-01-22 10:06:06.000000000 +0100
@@ -118,6 +118,10 @@
             fs.walk(basePath).on('data', item => {
                 //only watch directories
                 if (!item.stats.isDirectory()) return;
+
+                //don't watch the node_modules folder as it can be huge
+                if (item.path.indexOf(pathSlash + 'node_modules' + pathSlash, basePathLength) !== -1) return;
+
                 //don't watch any hidden folders or their children
                 if (item.path.indexOf(pathSlash + '.', basePathLength) !== -1) return;
                 this.watchers.push(fs.watch(item.path, { recursive: false }, this._fileChangeHandler.bind(this)));
