// stylelint/dist/rules/color-no-named/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (actual) {
  return function (root, result) {

    result.warn("'color-no-named' has been deprecated " + "and in 5.0 it will be removed. " + "Use 'color-named: \"never\"' instead.", {
      stylelintType: "deprecation",
      stylelintReference: "http://stylelint.io/user-guide/rules/color-named/"
    });

    var validOptions = (0, _utils.validateOptions)(result, ruleName, { actual: actual });
    if (!validOptions) {
      return;
    }

    root.walkDecls(function (decl) {
      (0, _postcssValueParser2.default)(decl.value).walk(function (node) {
        if ((0, _isCssColorName2.default)(node.value) && !node.quote) {
          (0, _utils.report)({
            message: messages.rejected(node.value),
            node: decl,
            index: (0, _utils.declarationValueIndexOffset)(decl) + node.sourceIndex,
            result: result,
            ruleName: ruleName
          });
        }
      });
    });
  };
};

var _postcssValueParser = require("postcss-value-parser");

var _postcssValueParser2 = _interopRequireDefault(_postcssValueParser);

var _isCssColorName = require("is-css-color-name");

var _isCssColorName2 = _interopRequireDefault(_isCssColorName);

var _utils = require("../../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ruleName = exports.ruleName = "color-no-named";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: function rejected(c) {
    return "Unexpected named color \"" + c + "\"";
  }
});
