// stylelint/dist/rules/declaration-block-no-single-line/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (actual) {
  return function (root, result) {

    result.warn("'declaration-block-no-single-line' has been deprecated " + "and in 5.0 it will be removed. " + "Use 'block-no-single-line' instead.", {
      stylelintType: "deprecation",
      stylelintReference: "http://stylelint.io/user-guide/rules/block-no-single-line/"
    });

    var validOptions = (0, _utils.validateOptions)(result, ruleName, { actual: actual });
    if (!validOptions) {
      return;
    }

    root.walkRules(function (rule) {

      if (!(0, _utils.isSingleLineString)((0, _utils.cssStatementBlockString)(rule))) {
        return;
      }

      (0, _utils.report)({
        message: messages.rejected,
        node: rule,
        index: (0, _utils.cssStatementStringBeforeBlock)(rule, { noBefore: true }).length,
        result: result,
        ruleName: ruleName
      });
    });
  };
};

var _utils = require("../../utils");

var ruleName = exports.ruleName = "declaration-block-no-single-line";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: "Unexpected single-line declaration block"
});
