// powerbi-visuals-package-validator/tasks/check-package.js
/*
 *  Power BI Visualizations
 *
 *  Copyright (c) Microsoft Corporation
 *  All rights reserved. 
 *  MIT License
 *
 *  Permission is hereby granted, free of charge, to any person obtaining a copy
 *  of this software and associated documentation files (the ""Software""), to deal
 *  in the Software without restriction, including without limitation the rights
 *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *  copies of the Software, and to permit persons to whom the Software is
 *  furnished to do so, subject to the following conditions:
 *
 *  The above copyright notice and this permission notice shall be included in 
 *  all copies or substantial portions of the Software.
 *
 *  THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
 *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
 *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
 *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
 *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 *  THE SOFTWARE.
 */

"use strict";

var gulp = require("gulp");
var using = require('gulp-using');
var path = require('path');
var fs = require('fs');
var gulputil = require('gulp-util');
var jsonschema = require('jsonschema');
var request = require('request');
var imagesize = require('image-size');

var pkglogger = undefined;
var resultResolver = undefined;
var urlPromiseCounter = 0;

var imageTypes = {
    icon: {
        title: 'icon',
        width: 20,
        height: 20
    },
    thumbnail: {
        title: 'thumbnail',
        width: 220,
        height: 176
    },
    screenshot: {
        title: 'screenshot',
        width: 424,
        height: 410
    }
};

var utils = {
    isVersion: function isVersion (version) {
        if (!version) {
            return false;
        }
        var pattern = new RegExp('^(?:0|[1-9][0-9]*)\.(?:0|[1-9][0-9]*)\.(?:0|[1-9][0-9]*)?$','i'); // 0.0.0 check
        return pattern.test(version)
    },

    isEmail: function isEmail (email) {
        if (!email) {
            return false;
        }
        var pattern = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
        return pattern.test(email);
    },

    isURL: function isURL (url) {
        if (!url) {
            return false;
        }
        var pattern = new RegExp('^(https?:\\/\\/)?'+ // protocol
        '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|'+ // domain name
        '((\\d{1,3}\\.){3}\\d{1,3}))'+ // OR ip (v4) address
        '(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*'+ // port and path
        '(\\?[;&a-z\\d%_.~+=-]*)?'+ // query string
        '(\\#[-a-z\\d_]*)?$','i'); // fragment locator
        return pattern.test(url);
    },

    isURLAvailable: function isURLAvailable (url) {
        request(url, function(error, response, body) {
            if(response && response.statusCode === 200) {
                if(body.length > 256) {
                    utils.log('ok', 'url ' + url, 'is available')
                } else {
                    utils.log('warning', 'url ' + url, 'seems to be empty')
                }
            } else if(response && response.statusCode !== 200) {
                utils.log('warning', 'url ' + url, 'answers ' + response.statusCode);
            } else {
                utils.log('error', 'url ' + url, 'is unavailable');
            }

            urlPromiseCounter--;
            if(!urlPromiseCounter) resultResolver(pkglogger);
        });
        urlPromiseCounter++;
    },

    isFileAvailable: function isFileAvailable (file_path) {
        try {
            fs.statSync(file_path);
            return true;
        } catch(err) {
            return !(err && err.code === 'ENOENT');
        }
    },

    isImage: function isImage (image) {
        if (!image) {
            return false;
        }

        return (/\.(gif|jpg|jpeg|tiff|png|svg)$/i).test(image);
    },

    isSVG: function isSVG (image) {
        if (!image) {
            return false;
        }

        return (/\.svg$/i).test(image);
    },

    isImageSize: function isImageSize(image_path, image_size) {
        var dimentions = imagesize(image_path);
        return (dimentions.width === image_size.width) &&
               (dimentions.height === image_size.height);

    },

    log: function log (type, field, message) {
        pkglogger.message.push({
            type: type,
            field: field,
            message:message,
        })

        pkglogger[type].push({
            field: field,
            text: message,
        })
    }
};

/**
 * Runs JSONSchema for given paths.
 * 
 * @param {String} path Path to files which will be linted.
 */
function checkPackage (packageJSONFile, isMSPackage, logger) {
    var packageJSONFile = packageJSONFile || "**/package.json";
    var packageDirectory = path.dirname(packageJSONFile);
    var packageJSON = JSON.parse(fs.readFileSync(packageJSONFile, "utf8"));

    pkglogger = logger.pkg

    // if new pbiviz tools generated
    var packageJSONSchema = '';
    if (packageJSON.metadata) {
        if(isMSPackage){
            packageJSONSchema = JSON.parse(fs.readFileSync('./config/package.pbiviztoolsgenerated.ms.schema.json', "utf8"))
        }else{
            packageJSONSchema = JSON.parse(fs.readFileSync(__dirname+'/../config/package.pbiviztoolsgenerated.schema.json', "utf8"))
        }
    } else {
        if(isMSPackage){
            packageJSONSchema = JSON.parse(fs.readFileSync('./config/package.ms.schema.json', "utf8")) 
        }else{
            packageJSONSchema = JSON.parse(fs.readFileSync(__dirname+'/../config/package.schema.json', "utf8"));
        }
    }

    var v = new jsonschema.Validator();

    // for custom field validation 
    // you could refactor ./config/package.schema.json with "pattern" field, 
    // but it will give undreadable messages like 
    // 'instance.version does not match pattern ^(?:0|[1-9][0-9]*)\\.(?:0|[1-9][0-9]*)\\.(?:0|[1-9][0-9]*)?$'
    // so lets do it in not flexible but readable way
    // with custom field "format" and its processor
    v.attributes.format = function formatProccessor(instance, schema, options, ctx) {
        if(typeof instance!='string') return;
        if(typeof schema.format!='string') throw new jsonschema.SchemaError('"format" expects a string', schema);

        //to support multiformat like "file|image|screenshot"
        var formats = schema.format.split('|');
        var i = 0;
        var format = '';

        for (i; i < formats.length; i++) {
            format = formats[i];
            switch(format) {
                case 'exact':
                    if ( instance !== schema.default ) {
                        return 'need to be equeal to ' + JSON.stringify(schema.default);
                    }
                    break;

                case 'x.x.x':
                    if (!utils.isVersion(instance)) {
                        return JSON.stringify(instance) + ' doesn`t much pattern ' + JSON.stringify(format);
                    }
                    break;

                case 'email':
                    if (!utils.isEmail(instance)) {
                        return JSON.stringify(instance) + ' is not an ' + JSON.stringify(format);
                    } 
                    break;

                case 'url':
                    if (!utils.isURL(instance)) {
                        return JSON.stringify(instance) + ' is not ' + JSON.stringify(format);
                    } else {
                        // async test for url
                        utils.isURLAvailable(instance);
                    }
                    break;

                case 'file':
                    if (!utils.isFileAvailable(path.join(packageDirectory, instance))) {
                        return JSON.stringify(instance) + ' is not in package';
                    }
                    break;

                case 'image':
                    if (!utils.isImage(instance)) {
                        return JSON.stringify(instance) + ' is not an ' + JSON.stringify(format);
                    }
                    break;

                case 'svg':
                    if (!utils.isSVG(instance)) {
                        return JSON.stringify(instance) + ' is not an ' + JSON.stringify(format);
                    }
                    break;

                case 'icon':
                case 'thumbnail':
                case 'screenshot':
                    if (!utils.isImageSize(path.join(packageDirectory, instance), imageTypes[format])) {
                        return JSON.stringify(instance) + ' as ' + 
                            JSON.stringify(format) + 
                            ' needs to be in size ' +
                            imageTypes[format].width + 'x' + imageTypes[format].height;
                    }
                    break;
            }
        };
    }

    var validationResult = v.validate(packageJSON, packageJSONSchema);
    var i = 0;

    if(validationResult.errors && validationResult.errors.length) {
        validationResult.errors.forEach( function (error, i) {
            utils.log('error', error.property, error.message);
        })
    }

    // Extra tests
    if (packageJSON && packageJSON.version && 
        packageJSON.visual && packageJSON.visual.version && 
        packageJSON.version !== packageJSON.visual.version
    ) {
        utils.log('error', 'instance.version and instance.visual.version', 'need to be the same')
    }

    var result = new Promise((resolve, reject) => {
        resultResolver = resolve
    });

    if(!urlPromiseCounter){
        resultResolver()
        return pkglogger;
    }else{
        return result;
    }

    // TODO try to use gulp pipeline to speed up
    // return gulp.src(path)
};

module.exports = checkPackage;
