// powerbi-visuals-package-validator/tasks/check-css.js
/*
 *  Power BI Visualizations
 *
 *  Copyright (c) Microsoft Corporation
 *  All rights reserved. 
 *  MIT License
 *
 *  Permission is hereby granted, free of charge, to any person obtaining a copy
 *  of this software and associated documentation files (the ""Software""), to deal
 *  in the Software without restriction, including without limitation the rights
 *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *  copies of the Software, and to permit persons to whom the Software is
 *  furnished to do so, subject to the following conditions:
 *
 *  The above copyright notice and this permission notice shall be included in 
 *  all copies or substantial portions of the Software.
 *
 *  THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
 *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
 *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
 *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
 *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 *  THE SOFTWARE.
 */

"use strict";

var gulp = require("gulp");
var gulpStylelint = require("gulp-stylelint").default;
var consoleReporter = require('gulp-stylelint-console-reporter').default;
var using = require('gulp-using');

var csslogger = undefined;

/**
 * Logs an issue to the console.
 * @param  {Object}    message Stylelint message.
 * @return {undefined}         Nothing is returned.
 */
function consoleMessageReporter(message) {
    csslogger.message.push(message);
    if(!csslogger[message.severity])
        csslogger[message.severity] = [];

    csslogger[message.severity].push({
        // line: message.line,
        // column: message.column,
        coords: 'line: '+message.line+', column: '+ message.column,
        text: message.text,
    })
}

/**
 * Logs a Stylelint result source and all its messages.
 * @param  {Object}    result Stylelint result.
 * @return {undefined}        Nothing is returned.
 */
/**
 * gulp-stylelint console reporter.
 * @module gulp-stylelint-console-reporter
 */

function consoleResultReporter(result) {
    result.messages.filter(function (message) {
        return message.plugin === 'stylelint';
    }).forEach(consoleMessageReporter);
}

/**
 * Reporter initialization function.
 * @return {Function} Reporter function.
 */
function myStylelintFormatter() {
    return function gulpStylelintConsoleReporter(resultList) {
        resultList.filter(function (result) {
            return result.messages.length > 0;
        }).forEach(consoleResultReporter);
    };
}

/**
 * Runs Styleslint linter for given paths.
 * 
 * @param {String} path Path to files which will be linted.
 */
function checkCSS (path, logger) {
    var path = path || "**/*.css";

    csslogger = logger.css;

    return gulp.src(path)
        // .pipe(using({prefix:'Checking', path:'relative', color:'green'}))
        .pipe(gulpStylelint({
            stylelint: {
                "extends": [__dirname+'/../config/.stylelintrc'],
                "legacyNumberedSeverities": false,
                "rules": {}
            },
            reporters: [
                myStylelintFormatter()
            ],
        }));
}

module.exports = checkCSS;
